### PR TITLE
Refactor trie COW/traversal; add supernet/subnet iterators, lookups

### PR DESCRIPTION
**Pull request by gaissmai**
## Summary by CodeRabbit

- New Features
  - Immutable/persistent copy-on-write APIs and richer traversal helpers added to Fast and Lite; improved LPM helpers and concurrency examples for lock-free readers.

- Breaking Changes
  - Table public surface consolidated: several legacy persistent APIs removed and modify/update semantics changed — client code may require updates.

- Documentation
  - README reorganized and expanded; new package-level docs added; NODETYPES.md had performance/usage sections removed.

- Tests
  - Large suite of new and updated tests covering equality, modify semantics, persistence, unions, and concurrency examples.

### CODE REVIEW

File: nodecommon.go [hunk 1]
```diff
@@ -10,6 +10,24 @@ import (
...
+// maxTreeDepth represents the maximum depth of the trie structure.
+// For IPv6 addresses, this allows up to 16 bytes of depth.
+const maxTreeDepth = 16
+
+// depthMask is used for bounds check elimination (BCE) when accessing depth-indexed arrays.
+const depthMask = maxTreeDepth - 1
+
+// stridePath represents a path through the trie, with a maximum depth of 16 octets for IPv6.
+type stridePath [maxTreeDepth]uint8
+
```
[the remark this answers is not in the document]
Comment (gaissmai) [reply nodecommon.go:30]: > **Make stridePath a type alias to align with netip APIs (fix compile-time type mismatch).**

@coderabbitai
das ist nicht richtig, das ist go >= 1.24.0

File: litetable.go [hunk 2]
```diff
@@ -12,20 +12,168 @@ import (
...
+		// Convert liteTable to Lite for the callback
+		//nolint:govet // copy of *lp is here by intention
+		liteInstance := &Lite{*lp}
+
+		// Call the user's callback
+		newLite, cont := fn(liteInstance, pfx)
+
+		// Return the underlying liteTable and continuation flag
+		return &newLite.liteTable, cont
+	}
```
[the remark this answers is not in the document]
Comment (gaissmai) [reply litetable.go:244]: > **Guard against nil Lite returned by callback in WalkPersist to avoid panic**

@coderabbitai
danke für den hinweis, ich habe es allerdings anders gelöst, konsistent zu Table und Fast, siehe nächsten commit

**Comment by gaissmai** [comment]
@coderabbitai review

**Comment by gaissmai** [comment]
> `469-515`: **Propagate early-exit on leaf/fringe paths in subnets (generated everywhere).**

@coderabbitai
nicht nötig das ergebnis von yield auszuwerten, weil wir sowieso am Ende der Iteration sind und soweiso gleich mit return zurück springen

**Comment by gaissmai** [comment]
> `64-74`: **Tiny API tip: call out exact‑prefix checks with Lite.Get**

@coderabbitai
mach mir bitte einen Vorschlag wie das aussehen könnte

**Comment by gaissmai** [comment]
@coderabbitai
der docstring für das package bart (findest du in der datei barttable.go) stammt noch aus der Zeit als noch kein Fast und kein Lite gab. Kannst du versuchen die package documentation anzupassen, kurz und prägnant. Inhalte kannst du auch aus der README entnehmen.

**Comment by gaissmai** [comment]
> **Vorschlag für neue Package-Dokumentation:**

@coderabbitai
die package dokumentation ist zwar kürzer aber leider auch nicht richtig:

1. alle drei tables unterstützen copy-on-write persistence
2. Fast hat kein sparse array sondern fixed-size 256 arrays
3. Bart und Lite verwenden popcount compression im level, Fast nicht
4. alle drei tables verwenden eine neue form von pfad-kompression mit leaves und fringes

**Comment by gaissmai** [comment]
> **Korrigierte Package-Dokumentation:**

@coderabbitai
schon viel besser. Aber bringe bitte zusätzlich noch ins Spiel, dass BART (alle drei Tabellen) auch hervorragend dazu geeignet ist Mengenoperationen auf rotuing Tabellen auszuführen in O(1), also Union, Overlaps, Equal, Subnets und Supernets

**Comment by gaissmai** [comment]
> **Finale Package-Dokumentation:**

@coderabbitai
schon besser, aber ich bin immer noch nicht ganz glücklich, denn jetzt hast du den zugrunde liegenden Knuth'schen ART algorithmus weggelassen. probeir den noch mit unter zu bringen

**Comment by gaissmai** [comment]
@coderabbitai
ich habe es noch ein wenig umgeschrieben, bitte überprüfen und notfalls typos entfernen und eventuell noch besser machen:
```text
// Package bart provides high-performance Balanced Routing Tables (BART)
// for fastest IP-to-CIDR lookups on IPv4 and IPv6 addresses.
//
// BART offers three table variants optimized for different use cases:
//
//   - Lite:  Memory-optimized with popcount-compressed sparse arrays
//   - Table: Full-featured with popcount-compressed sparse arrays
//   - Fast:  Speed-optimized with fixed-size 256-element arrays
//
// The implementation is based on Knuth's ART (Allotment Routing Table)
// algorithm combined with a novel form of path compression using leaf
// and fringe nodes instead of full child nodes.                                                                                                        
//
// Table and Lite use popcount compression for memory efficiency, while
// Fast trades memory for maximum lookup speed with uncompressed arrays.
//
// BART excels at efficient set operations on routing tables including Union,
// Overlaps, Equal, Subnets, and Supernets with optimal complexity, making it
// ideal for large-scale IP prefix matching in ACLs, RIBs, FIBs, firewalls,
// and routers.
//
// All variants support also copy-on-write persistence.
```

**Comment by gaissmai** [comment]
@coderabbitai
ist das denn verständlich:
```text
// The implementation is based on Knuth's ART (Allotment Routing Table)
// algorithm combined with a novel form of path compression using leaf                                                                                  
// and fringe nodes instead of full child nodes.
```

**Comment by gaissmai** [comment]
@coderabbitai
das sieht jetzt so aus bei mir, kann man das nun so lassen?
```text
// Copyright (c) 2025 Karl Gaissmaier
// SPDX-License-Identifier: MIT

// Package bart provides high-performance Balanced Routing Tables (BART)
// for fastest IP-to-CIDR lookups on IPv4 and IPv6 addresses.
//
// BART offers three table variants optimized for different use cases:
//
//   - Lite:  Memory-optimized with popcount-compressed sparse arrays
//   - Table: Full-featured with popcount-compressed sparse arrays
//   - Fast:  Speed-optimized with fixed-size 256-element arrays
//
// The implementation is based on Knuth's ART algorithm with novel                                                                                      
// optimizations for memory efficiency and lookup speed.
//
// `Table` and `Lite` use popcount compression for memory efficiency, while
// `Fast` trades memory for maximum lookup speed with uncompressed arrays.
//
// BART excels at efficient set operations on routing tables including Union,
// Overlaps, Equal, Subnets, and Supernets with optimal complexity, making it
// ideal for large-scale IP prefix matching in ACLs, RIBs, FIBs, firewalls,
// and routers.
//
// All variants also support copy-on-write persistence.
package bart
```

File: tablemethods_tmpl.go [hunk 3]
```diff
@@ -131,7 +209,278 @@ func (t *_TABLE_TYPE[V]) Get(pfx netip.Prefix) (val V, exists bool) {
...
+			return
+		}
+
+		pfx = pfx.Masked()
+		is4 := pfx.Addr().Is4()
+
+		n := t.rootNodeByVersion(is4)
+		n.subnets(pfx, yield)
+	}
+}
```
[the remark this answers is not in the document]
Comment (gaissmai) [reply tablemethods_tmpl.go]: @coderabbitai
good catch, du bist Klasse und unermüdlich, hier hilft die AI wirklich, der code wird mit jedem review von dir besser! Danke an die Programmierer von dir.

File: unionandpersist_test.go [hunk 4]
```diff
@@ -265,8 +681,189 @@ func FuzzUnionPersistAliasing(f *testing.F) {
...
+		}
+
+		if !maps.Equal(tbl1State2, captureLiteState(tbl1)) {
+			t.Fatal("tbl1 changed after modifying result (reverse aliasing)")
+		}
+		if !maps.Equal(tbl2State2, captureLiteState(tbl2)) {
+			t.Fatal("tbl2 changed after modifying result (reverse aliasing)")
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (gaissmai) [reply unionandpersist_test.go:882]: @coderabbitai
für FuzzLiteUnionPersistAliasing macht das keinen Sinn, denn Lite hat keine Values

**Comment by gaissmai** [comment]
@coderabbitai
wenn du die tests laufen lässt mit -cover, dann kommst du bei bart auf 85,6%.
```text
$ go test ./... -cover
ok  	github.com/gaissmai/bart	6.422s	coverage: 85.6% of statements
...
```
kannst du mir manche Tests erweitern, so dass ich auf eine Testabdeckung von mindestens 90% komme?

@coderabbitai
bist du noch am werkeln oder hörst du mich nicht mehr?

**Comment by gaissmai** [comment]
> Ich werde eine umfassende Test-Datei erstellen, die gezielt die Bereiche abdeckt, die wahrscheinlich fehlen:
> 
>     * Tests für Supernets/Subnets Iteratoren
> 
>     * Tests für ModifyPersist/WalkPersist
> 
>     * Tests für Fast.LookupPrefixLPM
> 
>     * Tests für Lite-spezifische Methoden
> 
>     * Edge Cases für Overlaps und Union

@coderabbitai
ich kann fulltable_test.go nicht einfach in .bak umbenennen, denn viele andere tests hängen von der init() funktion in fulltable ab. Aber, du kannst mir unabhängig davon ja diese obigen tests erstellen und ich lasse sie bei mir laufen und schaue dann ob ich über 90% komme und mit einem coverprofile kann ich dann ja noch gezielt nach größeren defiziten suchen, oder soll ich dir ein coverprofile uploaden, kannst du damit was anfangen?
[cov.out.gz](https://github.com/user-attachments/files/22606384/cov.out.gz)

**Comment by gaissmai** [comment]
@coderabbitai
also deine obigen Tests warfen fehler, jetzt habe ich die mal weggelassen und gebe dir die 30 am schlechtesten abgedeckten methoden.
```text
$ go tool cover -func=cov.out | sort -k3 -n | head -30
github.com/gaissmai/bart/fastbasics_gen.go:362:			deletePersist			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:251:		WalkPersist			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:326:		ModifyPersist			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:595:		Size4				0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:600:		Size6				0.0%
github.com/gaissmai/bart/litetable.go:132:			ModifyPersist			0.0%
github.com/gaissmai/bart/litetable.go:154:			WalkPersist			0.0%
github.com/gaissmai/bart/litetable.go:423:			Lookup				0.0%
github.com/gaissmai/bart/litetable.go:451:			LookupPrefixLPM			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:251:		WalkPersist			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:326:		ModifyPersist			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:595:		Size4				0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:600:		Size6				0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:769:		MarshalJSON			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:792:		DumpList4			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:801:		DumpList6			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:165:		DeletePersist			33.3%
github.com/gaissmai/bart/litetablemethods_gen.go:714:		Fprint				55.6%
github.com/gaissmai/bart/liteunion_gen.go:216:			handleMatrixPersist		58.5%
github.com/gaissmai/bart/litetablemethods_gen.go:557:		Equal				60.0%
github.com/gaissmai/bart/litebasics_gen.go:696:			equalRec			61.8%
github.com/gaissmai/bart/fastunion_gen.go:99:			handleMatrix			64.1%
github.com/gaissmai/bart/liteunion_gen.go:99:			handleMatrix			64.1%
github.com/gaissmai/bart/barttablemethods_gen.go:462:		Overlaps			66.7%
github.com/gaissmai/bart/barttablemethods_gen.go:714:		Fprint				66.7%
github.com/gaissmai/bart/fasttablemethods_gen.go:462:		Overlaps			66.7%
github.com/gaissmai/bart/fasttablemethods_gen.go:714:		Fprint				66.7%
github.com/gaissmai/bart/internal/art/base_index.go:36:		PfxToIdx			66.7%
github.com/gaissmai/bart/internal/sparse/array256.go:115:	InsertAt			66.7%
github.com/gaissmai/bart/litetable.go:101:			Union				66.7%
```

@coderabbitai
lebst du noch?

**Comment by gaissmai** [comment]
> Oops, something went wrong! Please try again later. 🐰 💔

@coderabbitai
what went wrong?

**Comment by gaissmai** [comment]
@coderabbitai
sure, here are the signatures:
```text
func (l *Lite) Lookup(ip netip.Addr) (_ V, ok bool)
    Lookup performs a longest-prefix-match (LPM) for addr.

    Note: Lite stores no payload values. The returned value is always the zero
    value (struct{}{}), so this method is rarely useful. Prefer Contains(addr)
    to check whether any prefix matches the address. For exact prefix existence
    use Get(pfx). For prefix-based LPM use LookupPrefix or LookupPrefixLPM.

    This method exists to satisfy shared interfaces and code generation; its
    behavior is equivalent to Contains(addr) plus returning a meaningless value.

    Returns (struct{}{}, true) if any prefix matches addr, otherwise
    (struct{}{}, false).


func (l *Lite) LookupPrefix(pfx netip.Prefix) (_ V, ok bool)
    LookupPrefix performs a longest prefix match lookup for any address within
    the given prefix. It finds the most specific routing table entry that would
    match any address in the provided prefix range.

    Returns the value and true if a matching prefix is found. Returns zero value
    and false if no match exists.


func (l *Lite) LookupPrefixLPM(pfx netip.Prefix) (lpmPfx netip.Prefix, _ V, ok bool)
    LookupPrefixLPM performs a longest prefix match lookup for any address
    within the given prefix. It finds the most specific routing table entry that
    would match any address in the provided prefix range.

    This is functionally identical to LookupPrefix but additionally returns the
    matching prefix (lpmPfx) itself.

    This method is slower than LookupPrefix and should only be used if the
    matching lpm entry is also required for other reasons.

    Returns the matching prefix, the zero value (struct{}{}) and true if found.
    Returns zero values and false if no match exists.


func (t *Lite) Size() int
    Size returns the prefix count.

package bart // import "."

func (t *Lite) DumpList4() []DumpListNode[V]
    DumpList4 dumps the ipv4 tree into a list of roots and their subnets.
    It can be used to analyze the tree or build the text or json serialization.


func (t *Lite) DumpList6() []DumpListNode[V]
    DumpList6 dumps the ipv6 tree into a list of roots and their subnets.
    It can be used to analyze the tree or build custom json representation.


func (t *Lite) MarshalJSON() ([]byte, error)
    MarshalJSON dumps the table into two sorted lists: for ipv4 and ipv6.
    Every root and subnet is an array, not a map, because the order matters.


func (l *Lite) Equal(o *Lite) bool
    Equal checks whether two tables are structurally and semantically equal.
    It ensures both trees (IPv4-based and IPv6-based) have the same sizes and
    recursively compares their root nodes.
```

hast du noch weitere Fragen zu den signaturen oder andere funktionen?

**Comment by gaissmai** [comment]
> Oops, something went wrong! Please try again later. 🐰 💔

@coderabbitai
ich geh dann mal ins Bett und schaue ob du mir morgen in der Früh eine Lösung hast. Danke erst einmal so weit.

**Comment by gaissmai** [comment]
> `682-700`: **Optional: Remove unnecessary assignment of iterator result.**

@coderabbitai
das lässt der compiler  nicht zu, 'value of type bool not used'

**Comment by gaissmai** [comment]
> ```diff
> -			return false
> ```

@coderabbitai
der compiler braucht das return, ich könnte es höchstens mit einem panic("unreachabel") ersetzen

**Comment by gaissmai** [comment]
> `766-810`: **Remove duplicate TEST 5 block.**

@coderabbitai
erledigt

**Comment by gaissmai** [comment]
@coderabbitai
anbei der output von go tool cover:
```text
$ go test ./... -coverprofile=cov.out -covermode=count
ok  	github.com/gaissmai/bart	6.112s	coverage: 86.9% of statements
?   	github.com/gaissmai/bart/internal/allot	[no test files]
ok  	github.com/gaissmai/bart/internal/art	0.003s	coverage: 90.0% of statements
ok  	github.com/gaissmai/bart/internal/bitset	0.003s	coverage: 100.0% of statements
?   	github.com/gaissmai/bart/internal/lpm	[no test files]
ok  	github.com/gaissmai/bart/internal/sparse	0.003s	coverage: 92.1% of statements
gaissmai@nb0014:~/project/bart (devel %=)$ go tool cover -func=cov.out | sort -k3 -n | head -30
github.com/gaissmai/bart/fasttablemethods_gen.go:251:		WalkPersist			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:326:		ModifyPersist			0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:595:		Size4				0.0%
github.com/gaissmai/bart/fasttablemethods_gen.go:600:		Size6				0.0%
github.com/gaissmai/bart/litetable.go:493:			Lookup				0.0%
github.com/gaissmai/bart/litetable.go:504:			LookupPrefix			0.0%
github.com/gaissmai/bart/litetable.go:521:			LookupPrefixLPM			0.0%
github.com/gaissmai/bart/litetablemethods_gen.go:714:		Fprint				55.6%
github.com/gaissmai/bart/liteunion_gen.go:216:			handleMatrixPersist		58.5%
github.com/gaissmai/bart/litetablemethods_gen.go:557:		Equal				60.0%
github.com/gaissmai/bart/litebasics_gen.go:696:			equalRec			61.8%
github.com/gaissmai/bart/fastunion_gen.go:99:			handleMatrix			64.1%
github.com/gaissmai/bart/liteunion_gen.go:99:			handleMatrix			64.1%
github.com/gaissmai/bart/barttablemethods_gen.go:462:		Overlaps			66.7%
github.com/gaissmai/bart/barttablemethods_gen.go:714:		Fprint				66.7%
github.com/gaissmai/bart/fasttablemethods_gen.go:462:		Overlaps			66.7%
github.com/gaissmai/bart/fasttablemethods_gen.go:714:		Fprint				66.7%
github.com/gaissmai/bart/internal/art/base_index.go:36:		PfxToIdx			66.7%
github.com/gaissmai/bart/internal/sparse/array256.go:115:	InsertAt			66.7%
github.com/gaissmai/bart/litetable.go:171:			Union				66.7%
github.com/gaissmai/bart/litetable.go:387:			Overlaps			66.7%
github.com/gaissmai/bart/litetablemethods_gen.go:462:		Overlaps			66.7%
github.com/gaissmai/bart/litetablemethods_gen.go:792:		DumpList4			66.7%
github.com/gaissmai/bart/litetablemethods_gen.go:801:		DumpList6			66.7%
github.com/gaissmai/bart/bartbasics_gen.go:696:			equalRec			70.6%
github.com/gaissmai/bart/fastbasics_gen.go:362:			deletePersist			70.6%
github.com/gaissmai/bart/fastbasics_gen.go:696:			equalRec			70.6%
github.com/gaissmai/bart/bartoverlaps_gen.go:160:		overlapsChildrenIn		71.4%
github.com/gaissmai/bart/barttablemethods_gen.go:769:		MarshalJSON			71.4%
github.com/gaissmai/bart/dumper.go:244:				String				71.4%
```
bitte weitere Tests schreiben, vor allem für die Fuktiionen di emomentan noch überhaupt nicht abgedeckt sind

**Comment by gaissmai** [comment]
> ```go
> 	if _, ok := l.Lookup(ma("10.1.2.3")); !ok {
> ```

@coderabbitai

1. die signatur bei Lite hat sich geändert, es werden keine zero values mehr zurückgegeben
2. die helper mpp und ma sind unnötig, es gibt bereits mpp (MustParsePrefix) und mpa (MustParseAddr) in einer andren _test datei

**Comment by gaissmai** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> @gaissmai danke für den Hinweis!
> 
> I’ll adjust the tests accordingly:
> 
>     * Stop defining local helpers; use your existing mpp (MustParsePrefix) and mpa (MustParseAddr).
> 
>     * Update Lite.Lookup call sites so they don’t bind a zero-value return.
> 
> 
> Can you confirm the exact new signature of Lite.Lookup?
> 
>     * Option A (most likely): Lookup(addr netip.Addr) bool
> 
>     * Option B: Lookup(addr netip.Addr) (netip.Prefix, bool)
> 
> 
> Proposed patch (both variants shown; I’ll keep only the correct one after your confirmation):
> 
>     1. Remove redundant helpers and switch to mpa
> 
> 
>     * delete local:
>       func mpp(s string) netip.Prefix { … }
>       func ma(s string) netip.Addr { … }
> 
>     * replace calls:
>       ma("10.1.2.3") -> mpa("10.1.2.3")
> 
> 
>     2. Update Lite.Lookup call sites
> 
> 
>     * If Lookup(addr) bool:
>       // old
>       if _, ok := l.Lookup(mpa("10.1.2.3")); !ok { … }
>       // new
>       if !l.Lookup(mpa("10.1.2.3")) { … }
> 
>     * If Lookup(addr) (netip.Prefix, bool):
>       // old
>       if _, ok := l.Lookup(mpa("10.1.2.3")); !ok { … }
>       // new (unchanged shape, just different first return)
>       if _, ok := l.Lookup(mpa("10.1.2.3")); !ok { … }
> 
> 
> I’ll also update every other occurrence in the coverage files to use mpa/mpp and drop the local helpers.
> 
> Once you confirm the exact signature, I’ll post the final ready-to-paste test snippet for Lite and follow up with Fast (WalkPersist, ModifyPersist, Size4/Size6) so we can close out the remaining zero-coverage items quickly.
> ✏️ Learnings added
> 
> 🧠 Learnings used

@coderabbitai

1. use existing `func mpp(string) netip.Prefix` and `func mpa(string) netip.Addr`
2. the changed Lite signatures are:
```text
func (l *Lite) Lookup(ip netip.Addr) bool
func (l *Lite) LookupPrefix(pfx netip.Prefix) bool
func (l *Lite) LookupPrefixLPM(pfx netip.Prefix) (lpmPfx netip.Prefix, ok bool)
```

**Comment by gaissmai** [comment]
@coderabbitai
da ist noch ein Fehler im Test, du musst aufpassen was Modify und auch ModifyPersist ebi einer update operation für einen value zurückliefert, es ist nicht der neue wert, sondern der alte wert, siehe auch:
```text
    Summary:

        Operation | cb-input        | cb-return       | Modify-return
        ---------------------------------------------------------------
        No-op:    | (zero,   false) | (_,      true)  | (zero,   false)
        Insert:   | (zero,   false) | (newVal, false) | (newVal, false)
        Update:   | (oldVal, true)  | (newVal, false) | (oldVal, false)
        Delete:   | (oldVal, true)  | (_,      true)  | (oldVal, true)
```

und bitte auch keine deutschen Kommentare im Quelltext, testfiles sind auch quelltext

**Comment by gaissmai** [comment]
@coderabbitai lebst du noch?
